src/notifications.c

```diff
@@ -350,28 +409,35 @@ void ClusterSlotMigrationEvent(RedisModuleCtx *ctx, RedisModuleEvent eid, uint64
 
     case REDISMODULE_SUBEVENT_CLUSTER_SLOT_MIGRATION_IMPORT_STARTED:
       RedisModule_Log(RSDummyContext, "notice", "Got ASM import started event.");
-      in_asm_import = true;
-      should_filter_slots = true;
       ASM_StateMachine_StartImport(slots);
       workersThreadPool_OnEventStart();
       break;
     case REDISMODULE_SUBEVENT_CLUSTER_SLOT_MIGRATION_IMPORT_COMPLETED:
       ASM_StateMachine_CompleteImport(slots);
-      Slots_DropCachedLocalSlots(); // Local slots have changed, drop the cache
     case REDISMODULE_SUBEVENT_CLUSTER_SLOT_MIGRATION_IMPORT_FAILED:
       RedisModule_Log(RSDummyContext, "notice", "Got ASM import %s event.", subevent == REDISMODULE_SUBEVENT_CLUSTER_SLOT_MIGRATION_IMPORT_FAILED ? "failed" : "completed");
-      in_asm_import = false;
-      should_filter_slots = in_asm_trim;
       // Since importing is done in a part-time job while redis is running other commands, we notify
       // the thread pool to no longer receive new jobs, and terminate the threads ONCE ALL PENDING JOBS ARE DONE.
       workersThreadPool_OnEventEnd(false);
+      if (!IsEnterprise() && subevent == REDISMODULE_SUBEVENT_CLUSTER_SLOT_MIGRATION_IMPORT_COMPLETED) {
+        RedisTopologyUpdater_StopAndRescheduleInmediately(ctx);
+      }
       break;
 
     // case REDISMODULE_SUBEVENT_CLUSTER_SLOT_MIGRATION_MIGRATE_STARTED:
     // case REDISMODULE_SUBEVENT_CLUSTER_SLOT_MIGRATION_MIGRATE_FAILED:
     case REDISMODULE_SUBEVENT_CLUSTER_SLOT_MIGRATION_MIGRATE_COMPLETED:
       ASM_StateMachine_CompleteMigration(slots);
-      Slots_DropCachedLocalSlots(); // Local slots have changed, drop the cache
+      // Start 2 timers. One for the minimal delay, and one for the maximal delay.
+      RedisModule_Log(ctx, "notice", "Got ASM migrate completed event.");
+      RedisModule_ClusterDisableTrim(ctx);
+      trimmingDelayCtx.checkTrimmingStateTimerId = RedisModule_CreateTimer(ctx, RSGlobalConfig.minTrimDelayMS, checkTrimmingStateCallback, slots);
+      trimmingDelayCtx.enableTrimmingTimerId = RedisModule_CreateTimer(ctx, RSGlobalConfig.maxTrimDelayMS, enableTrimmingCallback, slots);
+      trimmingDelayCtx.checkTrimmingStateTimerIdScheduled = true;
+      trimmingDelayCtx.enableTrimmingTimerIdScheduled = true;
+      if (!IsEnterprise()) {
+        RedisTopologyUpdater_StopAndRescheduleInmediately(ctx);
+      }

```

Comment:
Crucial point I missed.

Since redis is shipped with RQE by default, we should avoid delaying the trimming if the module is not being used.

I think the simplest check to make is whether we currently have any indexes. If not, we won't need the delay.
Also make sure not to enable the trimming if we didn't delay it (I think not creating the timers in this case will ensure that)

Reply:
Oh, SUPER IMPORTANT! Great catch!